pkg/core/storage/memcached_store.go

```diff
@@ -77,21 +83,120 @@ func (s *MemCachedStore) GetBatch() *MemBatch {
 
 // Seek implements the Store interface.
 func (s *MemCachedStore) Seek(key []byte, f func(k, v []byte)) {
+	// Create memory store snapshot not to hold the lock.
+	memSnapshot := NewMemoryStore()
 	s.mut.RLock()
-	defer s.mut.RUnlock()
-	s.MemoryStore.seek(key, f)
-	s.ps.Seek(key, func(k, v []byte) {
-		elem := string(k)
-		// If it's in mem, we already called f() for it in MemoryStore.Seek().
-		_, present := s.mem[elem]
-		if !present {
-			// If it's in del, we shouldn't be calling f() anyway.
-			_, present = s.del[elem]
+	for k, v := range s.MemoryStore.mem {
+		memSnapshot.mem[k] = v
+	}
+	for k, v := range s.MemoryStore.del {
+		memSnapshot.del[k] = v
+	}
+	s.mut.RUnlock()
+
+	var (
+		data1Cont = make(chan bool)
+		data2Cont = make(chan bool)
```

Comment:
These channels are not really needed, initially you read from `data1` and `data2`, then only from one of them.